### PR TITLE
Stabilize snapping 2D transforms to pixel

### LINKED ISSUE #56793: Vulkan: 2D Pixel Snap not working
### Godot version

4.0dev (ce6de71)

### System information

Windows 10, NVIDIA GeForce GTX 960

### Issue description

I've seen other issues like this but they're all closed as fixed. I'm still having this issue. Pixels are not snapping correctly with either or both pixel snap options.
1 pixel gap in between collisions.

### Steps to reproduce

Create a project with a floor and characterbody2d with basic movement functions, gravity, etc.
Set display stretch mode to `viewport` and `keep`.
Enable `Snap 2d Transforms to Pixel` or `Snap 2d Vertices to Pixel` or both.

image

### Minimal reproduction project

[PixelSnapBug.zip](https://github.com/godotengine/godot/files/7872520/PixelSnapBug.zip)

### LINKED ISSUE #84632: Sprite "crunch" when placed in certain positions on screen
### Godot version

v4.2.beta5.official [4c96e9676]

### System information

Godot v4.2.beta5 - Windows 10.0.22621 - Vulkan (Forward+) - dedicated AMD Radeon RX 7900 XTX (Advanced Micro Devices, Inc.; 31.0.22023.1014) - AMD Ryzen 9 7950X3D 16-Core Processor (32 Threads)

### Issue description

In 4.2, there is an issue mostly visible on pixel art games with small resolutions that scale up to larger screens. While the game is running, moving sprites to certain places in the screen will cause the sprites to appear uneven or "crunched", and it's usually more noticeable if the sprite is animated.

example

Window settings:
image

This seems to happen when moving the position of the sprite to X.5.
This happens in all 3 renderers.
This issue does not occur in Godot 4.1.3.
This issue occurs as early as 4.2 dev1.

Note: I'm using a 4K monitor with Windows scaled up to 175%.

### Steps to reproduce

1. Download the reproduction project
2. Open the project
3. Run the game
4. If the issue doesn't immediately show itself, slightly moving the character to the left/right a few times should trigger it.

### Minimal reproduction project

[SpriteCrunchIssue.zip](https://github.com/godotengine/godot/files/13301032/SpriteCrunchIssue.zip)

### PR DESCRIPTION
**Pull request by adamscott**
# tl;dr
This PR simplifies greatly the snapping of 2D transforms to pixel. Instead of relying on flooring sprites, CanvasItems (so twice flooring for sprites) and the camera transform, this PR only rounds each CanvasItem position.

# Co-authorship
This PR is heavily inspired by #84380 by @KeyboardDanni, so I added them to the commit as co-author.

# Why?
There's some issues with pixel perfect both due to:
- the numerous places where flooring occurs;
- the flooring itself;

# Superseding
- Supersedes #84380
- Supersedes #87058

# Tests
- Passes #84380 [PixelPerfectTest.zip](https://github.com/godotengine/godot/files/13244300/PixelPerfectTest.zip)

### master

[master.webm](https://github.com/godotengine/godot/assets/270928/d95512f2-af68-4cb8-8737-82bb01147f6b)

### This PR

[87297.webm](https://github.com/godotengine/godot/assets/270928/f626b3d2-2a7f-4388-9b84-4a918f40b40f)

## [Merlin: Scale of the Magic](https://github.com/nicolasbize/Scalazard)

(author note: the project was edited to add debug info)

### master

[master.webm](https://github.com/godotengine/godot/assets/270928/9d970751-83e9-458c-9639-b7e9a8586aca)

### This PR

[87297.webm](https://github.com/godotengine/godot/assets/270928/0d8b030c-e266-4258-9b7a-4ae61c20641f)

_Production edit: closes godotengine/godot-roadmap#79_

*Bugsquad edit:*
- Fixes https://github.com/godotengine/godot/issues/56793
- Fixes https://github.com/godotengine/godot/issues/84632

### CODE REVIEW

**Comment by adamscott** [comment]
Just realized that it doesn't completely pass the #84380 pixel perfect test. This PR has some scaling/rotation issues.

**Comment by YuriSizov** [comment]
Could you please rebase on the current master? We had an issue with macOS builds which failed your CI.

**Comment by markdibarry** [comment]
Just tested with both pixel perfect test projects, and both look as expected.

**Review by akien-mga — APPROVED** [review]
Great research everyone!

Let's give this a spin and see how pixel art users like it in the next dev snapshot (4.3-dev4).

**Comment by akien-mga** [comment]
Thanks!

**Comment by alvinhochun** [comment]
A bit late to the party, but I have the feeling that instead of `round`ing, it should have been `floor(x + 0.5)` instead. This is because rounding to nearest has different behaviour for positive and negative values when the fractional part is `0.5`.

Consider this: We position a sprite at x=0.5, which is snapped to x=1 on screen. We move it to the left by 1px. Where should the sprite be at on screen?

The expectation should be x=0. However, since the subpixel position is now x=-0.5, the sprite is actually snapped to x=-1 on screen thanks to rounding. Which means the supposedly 1px move becomes a 2px move on screen.
